deepcell/graph_utils.py

```diff
@@ -0,0 +1,274 @@
+
+    # Instantiate the adjacency matrix
+    adjacency_matrix = np.zeros((label_image.shape[0], max_cells, max_cells))
+    centroid_matrix = np.zeros((label_image.shape[0], max_cells, 2))
+    label_matrix = np.zeros((label_image.shape[0], max_cells))
+
+    for batch in range(label_image.shape[0]):
+        label_image_batch = label_image[batch]
+
+        props = regionprops_table(label_image_batch, properties=['centroid', 'label'])
+        centroids = np.stack([props['centroid-0'], props['centroid-1']], axis=-1)
+        labels = props['label']
+        distances = cdist(centroids, centroids, metric='euclidean')
+
+        threshold = distances.mean()*distance_threshold
+
+        adjacency_matrix[batch,
+                         0:distances.shape[0],
+                         0:distances.shape[1]] = distances < threshold
+
+        if not self_connection:
+            adjacency_matrix[batch,
+                             0:distances.shape[0],
+                             0:distances.shape[1]] -= np.eye(distances.shape[0])
+
+        centroid_matrix[batch, 0:centroids.shape[0], :] = centroids
+        label_matrix[batch, 0:labels.shape[0]] = labels
+
+    return adjacency_matrix, centroid_matrix, label_matrix
+
+
+def get_cell_features(label_image, image):
+    """Extract feature vectors from cells in multiplexed imaging data
+
+    Args:
+        image (float): Multiplexed imaging dataset
+        label_image (int): Label image where every cell has been given a
+            unique integer id
+
+    Raises:
+        ValueError: if image and label_image batch size does not match
+
+    Returns:
+        np.array: matrix consisting of feature vectors for each cell
+    """
+    if image.shape[0] != label_image.shape[0]:
+        raise ValueError("image and label_image must have the same size for "
+                         "the batch dimmension")
+
+    # Find out the max number of cells
+    max_cells = get_max_cells(label_image)
+
+    # Instantiate the feature matrix
+    feature_matrix = np.zeros((image.shape[0], max_cells, image.shape[-1]))
+    label_matrix = np.zeros((label_image.shape[0], max_cells))
+
+    # Extract the features
+    # TODO: Add flag to include morphology features
+    for batch in range(image.shape[0]):
+        image_batch = image[batch]
+        label_image_batch = np.squeeze(label_image[batch])
+        for feature in range(image.shape[-1]):
+            props = regionprops_table(label_image_batch,
+                                      intensity_image=image_batch[..., feature],
+                                      properties=['mean_intensity', 'label'])
+
+            mean_intensity = props['mean_intensity']
+            labels = props['label']
+
+            feature_matrix[batch, 0:mean_intensity.shape[0], feature] = mean_intensity
+            label_matrix[batch, 0:labels.shape[0]] = labels
+
+    return feature_matrix
+
+
+def get_celltypes(label_image, celltype_image):
```

Comment:
Let's refactor `celltype_image` to `classification_image` or `semantic_class_image` or something that is not specific to cell types.